File: clojure/src/project_euler/sol/p0080.clj
(ns project-euler.sol.p0080
  (:require
   [project-euler.lib.math :as my-math]
   [project-euler.lib.util :as util]))

(defn solve
  ([]
   (solve 100 100))
  ([upper n-digit]
   (let [pow10 (my-math/pow 10 (* (dec n-digit) 2))
         xf (comp (remove #(my-math/square? %))
                  (map #(my-math/isqrt (* % pow10))) ; Use math/isqrt due to bigint value.
                  (map #(util/digits %))
                  (map #(take n-digit (reverse %)))
                  (map #(apply + %)))]
     (transduce xf + (range 1 (inc upper))))))

File: haskell/src/Sol/P0080.hs
module Sol.P0080 (compute, solve) where

import Mylib.Math (isqrt)
import Mylib.Util (digits)

compute :: Integer -> Int -> String
compute limit ndigit =
    show $
        sum
            [ sum . take ndigit . reverse $ digits x
            | n <- [1 .. limit]
            , isqrt n * isqrt n /= n
            , let x = isqrt (pow10 * n)
            ]
  where
    pow10 = 10 ^ ((ndigit - 1) * 2) :: Integer

solve :: String
solve = compute 100 100
